File: python/SSAFY 입과전/mid03.py

def plusShape(board, x, y, n, m) :
    total = 0;
    total += sum(board[x][max(0, y - m + 1) : min(n, y + m)]) # 가로 행 더하기 
    for i in range(max(0, x - m + 1), min(n, x + m)):
        total += board[i][y]
    total -= board[x][y]; 
    return total; 
     
def xShape(board, x, y, n, m) :
    total = board[x][y];
    for i in range(1, m):
        if x + i >= n or x - i < 0 or y + i >= n or y - i < 0:
            continue; 
        total += board[x + i][y + i] + board[x - i][y + i] + board[x + i][y - i] + board[x - i][y - i]
    return total; 


T = int(input())

for test_case in range(1, T + 1): 
    n, m = map(int, input().split(" "))
    board = [] 
    for i in range(n): 
        board.append(list(map(int, input().split()))); 
    maxValue = 0;

    for i in range(n):
        for j in range(n): 
            plus = plusShape(board, i, j, n, m)
            multi = xShape(board, i, j, n, m)
            maxValue = max(maxValue, plus, multi)
    print(f"#{test_case} {maxValue}")

File: python/SSAFY 입과전/Mid03.java

import java.util.Scanner;


class Solution
{
	public static void main(String args[]) throws Exception
	{
		
		Scanner sc = new Scanner(System.in);
		int T=sc.nextInt();

		for(int test_case = 1; test_case <= T; test_case++){
            int n = sc.nextInt(); 
            int m = sc.nextInt(); 

            int[][] board = new int[n][n];
            for(int i = 0;i < n; i++) {
                for(int j = 0; j < n; j++) {
                    board[i][j] = sc.nextInt(); 
                }
            }

            int max = 0; 

            // + 모양으로 뿌리기
            for(int i = 0;i < n; i++) {
                for(int j = 0; j < n; j++) {
                    int total = board[i][j]; 
                    for(int k = 1; k < m; k++) {
                        if(j + k < n) total += board[i][j + k]; 
                        if(j - k >= 0) total += board[i][j - k]; 
                        if(i + k < n) total += board[i + k][j]; 
                        if(i - k >= 0) total += board[i - k][j]; 
                    }
                    max = Math.max(max, total); 
                }
            }

            // x 모양으로 뿌리기 
            for(int i = 0;i < n; i++) {
                for(int j = 0; j < n; j++) {
                   int total = board[i][j]; 
                   for(int k = 1; k < m; k++) {
                        if(i - k >= 0 && j - k >= 0) total += board[i - k][j - k]; 
                        if(i - k >= 0 && j + k < n) total += board[i - k][j + k]; 
                        if(i + k < n && j - k >= 0) total += board[i + k][j - k]; 
                        if(i + k < n && j + k < n) total += board[i + k][j + k]; 
                   }
                   max = Math.max(max, total); 
                }
            }

            System.out.printf("#%d ", test_case);
            System.out.println(max);
		}
	}
}